### DISCUSSION TITLE
Spindle synchronization and MPG requirements

### DISCUSSION THREAD

**Discussion by Krapovnitsky** [grblHAL/iMXRT1062#44] [General]
HI! Trying to understand if MPG and spindle synchronization possible within single configuration on iMXRT1062 grblHAL latest release. If no, what can be done to overcome the limitation? I consider all possible options. Thank you.
Ed.

**Comment by terjeio** [comment]
It depends on what you mean by a single configuration.
Spindle sync can be enabled by a [single compile time configuration](https://github.com/grblHAL/iMXRT1062/blob/2d19f2af392fcfe8d0f07a85f7f038876a0102b4/grblHAL_Teensy4/src/my_machine.h#L78-L79) provided that the map file used has the neccessary pin definitions.
There are several run time configurations needed: lathe mode, spindle PPR (pulses per revolution) and PID parameters for the control loop. MPG mode is enabled by [one or two compile time parameters](https://github.com/grblHAL/iMXRT1062/blob/2d19f2af392fcfe8d0f07a85f7f038876a0102b4/grblHAL_Teensy4/src/my_machine.h#L65-L68) and can be enabled at the same time as spindle sync provided that the pins used are different from those used by spindle sync.

**Reply by Krapovnitsky** [reply]
Thank you for the response!
I understand now that GRBL MPG DRO BoosterPack project is already supported by iMXRT1062 grblHAL. Seems like repeating the MPG kit by others is problematic because of lack of documentation or kits, especially if RP2040 MCU used (no pinout available).
- [x]  As far as I understood some sort of serial connection used between RP2040 and Teensy 4.1. Is that a serial connection over USB or over GPIO pins?

Is that possible to use just RP2040 without screen and MSP430 for MPG function?

Regards,
 Ed.

**Reply by terjeio** [reply]
> No pinout available

RP2040 pins are defined [here](https://github.com/terjeio/GRBL_MPG_DRO_BoosterPack/blob/master/Firmware/RP2040/hal_rp2040/driver.h) and [design files are found here](https://github.com/terjeio/GRBL_MPG_DRO_BoosterPack/tree/master/Firmware) (KiCad format). I believe I have two bare PCBs available and I can supply a preprogrammed MSP340. I have no plans to produce kits, I'll leave that to somebody else.

>  As far as I understood some sort of serial connection used between RP2040 and Teensy 4.1. Is that a serial connection over USB or over GPIO pins?

GPIO. USB needs host mode support added to the iMXRT1062 driver.

> Is that possible to use just RP2040 without screen and MSP430 for MPG function?

Not without code changes if using my firmware, if using your own then yes.

An example of a third party MPG is the [I2C Jog Controller](https://github.com/Expatria-Technologies/I2C_Jog_Controller).
